### ISSUE TITLE
Freezing a model with Exponential Moving Average gives different probabilities

### ISSUE THREAD

**Issue by jumpyknight** [tensorflow/tensorflow#25912]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution: Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 1.9.0
- Python version: 2.7.12
- Bazel version (if compiling from source): 0.15.0
- GCC/Compiler version (if compiling from source): 5.4.0
- CUDA/cuDNN version: 9.0.176
- GPU model and memory:

I'm trying to freeze a model using Exponential Moving Average (EMA) variables for training and inference.

The freezing and inference code work normally, but the inferred probabilities are slightly different (about 0.05) with the ones from the original code. These results become identical (differnce < 1e-6) if I turn off the EMA in both models.

Am I doing the freezing incorrectly, or the freeze_graph tool is just unable to treat EMA?

Code I am using for freezing graph:

    from __future__ import print_function
    import tensorflow as tf
    from nets.inception_v3 import inception_v3, inception_v3_arg_scope
    from tensorflow.python.framework import graph_util
    import sys 
    slim = tf.contrib.slim

    checkpoint_file = '/my/model'

    with tf.Graph().as_default() as graph:

        images = tf.placeholder(shape=[None, 100, 221, 6], dtype=tf.float32, name = 'input')

        with slim.arg_scope(inception_v3_arg_scope()):
            logits, end_points = inception_v3(images, num_classes = 3, create_aux_logits = False, is_training = False)

        variables_to_restore = slim.get_variables_to_restore()

        MOVING_AVERAGE_DECAY = 0.9999
        variable_averages = tf.train.ExponentialMovingAverage(
            MOVING_AVERAGE_DECAY)
        for var in variables_to_restore:
            tf.add_to_collection(tf.GraphKeys.MOVING_AVERAGE_VARIABLES, var)
        variables_to_restore = variable_averages.variables_to_restore()        #This line is commented if EMA is turned off

        saver = tf.train.Saver(variables_to_restore)

        #Setup graph def
        input_graph_def = graph.as_graph_def()
        output_node_names = "InceptionV3/Predictions/Reshape_1"
        output_graph_name = "./frozen_inception_v3_new_100_221_ema.pb"

        with tf.Session() as sess:
            saver.restore(sess, checkpoint_file)

            #Exporting the graph
            print ("Exporting graph...")
            output_graph_def = graph_util.convert_variables_to_constants(
                    sess,
                    input_graph_def,
                    output_node_names.split(","))

            with tf.gfile.GFile(output_graph_name, "wb") as f:
                f.write(output_graph_def.SerializeToString())

This code is modified from https://gist.github.com/kwotsin/8e43f5db4815e1f1af37da70d0933d8b . The EMA part is the same as the original code.

**Comment by jumpyknight** [comment]
The problem is solved.
The EMA part in the code I use is not correct.
    MOVING_AVERAGE_DECAY = 0.9999
    variable_averages = tf.train.ExponentialMovingAverage(
        MOVING_AVERAGE_DECAY)
    for var in variables_to_restore:
        tf.add_to_collection(tf.GraphKeys.MOVING_AVERAGE_VARIABLES, var)
    variables_to_restore = variable_averages.variables_to_restore()
If I remove
    for var in variables_to_restore:
        tf.add_to_collection(tf.GraphKeys.MOVING_AVERAGE_VARIABLES, var)
The results become the same with the original code.
